lnwallet/rpcwallet/remote_signer_client.go

```diff
@@ -0,0 +1,977 @@
+// new connections.
+//
+// NOTE: This is part of the SignCoordinatorStreamFeeder interface.
+func (s *StreamFeeder) Stop() {
+	close(s.quit)
+
+	s.wg.Wait()
+}
+
+// GetStream returns a new stream to the watch-only node, by making an
+// outbound gRPC connection to the watch-only node. The function also returns a
+// cleanup function that closes the connection, which should be called when the
+// stream is no longer needed.
+//
+// NOTE: This is part of the SignCoordinatorStreamFeeder interface.
+func (s *StreamFeeder) GetStream(streamCtx context.Context) (
+	StreamClient, func(), error) {
+
+	select {
+	// Don't run if the StreamFeeder has already been shutdown.
+	case <-s.quit:
+		return nil, nil, ErrShuttingDown
+	default:
+	}
+
+	// Create a new outbound gRPC connection to the watch-only node.
+	conn, err := s.getClientConn()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	cleanUp := func() {
+		conn.Close()
+	}
+
+	// Wrap the connection in a WalletKitClient.
+	walletKitClient := walletrpc.NewWalletKitClient(conn)
+
+	// Create a new stream to the watch-only node.
+	stream, err := walletKitClient.SignCoordinatorStreams(streamCtx)
+	if err != nil {
+		cleanUp()
+
+		return nil, nil, err
+	}
+
+	return stream, cleanUp, nil
+}
+
+// getClientConn creates a new outbound gRPC connection to the watch-only node.
+func (s *StreamFeeder) getClientConn() (*grpc.ClientConn, error) {
+	// If we fail to connect to the watch-only node within the
+	// configured timeout we should return an error.
+	ctxt, cancel := context.WithTimeout(
+		context.Background(), s.timeout,
+	)
+	defer cancel()
+
+	// Load the specified macaroon file for the watch-only node.
+	macBytes, err := os.ReadFile(s.macaroonPath)
+	if err != nil {
+		return nil, fmt.Errorf("could not read macaroon file: %w", err)
+	}
+
+	mac := &macaroon.Macaroon{}
+
+	err = mac.UnmarshalBinary(macBytes)
+	if err != nil {
+		return nil, fmt.Errorf("could not unmarshal macaroon: %w", err)
+	}
+
+	macCred, err := macaroons.NewMacaroonCredential(mac)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"could not create macaroon credential: %w", err)
+	}
+
+	// Load the specified TLS cert for the watch-only node.
+	tlsCreds, err := credentials.NewClientTLSFromFile(s.tlsCertPath, "")
+	if err != nil {
+		return nil, fmt.Errorf("could not load TLS cert: %w", err)
+	}
+
+	opts := []grpc.DialOption{
+		grpc.WithBlock(),
+		grpc.WithTransportCredentials(tlsCreds),
+		grpc.WithPerRPCCredentials(macCred),
+	}
+
+	var (
+		// A channel to signal when has successfully been created.
+		connDoneChan = make(chan *grpc.ClientConn)
+		errChan      = make(chan error)
+	)
+
+	// Now let's try to connect to the watch-only node. We'll do this in a
+	// goroutine to ensure we can exit if the quit channel is closed. If the
+	// quit channel is closed, the context will also be canceled, hence
+	// stopping the goroutine.
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+
```

Comment:
I think we can simplify the readability of this method quite a bit if we add a helper like this to `fn`:

```
func CancelOrQuit(ctx context.Context, quit chan struct{}) (context.Context,
	context.CancelFunc) {

	ctx, cancel := context.WithCancel(ctx)
	go func() {
		select {
		case <-quit:
			cancel()
		case <-ctx.Done():
		}
	}()

	return ctx, cancel
}
```

Then at the top of this func you just have:

```
// If we fail to connect to the watch-only node within the
	// configured timeout we should return an error.
	ctxt, cancel := context.WithTimeout(ctx, s.timeout)
	defer cancel()
	ctx, _ = CancelOrQuit(ctxt, s.quit)
```

and then you can just:
```
opts := []grpc.DialOption{
		grpc.WithBlock(),
		grpc.WithTransportCredentials(tlsCreds),
		grpc.WithPerRPCCredentials(macCred),
	}

	log.Infof("Attempting to connect to the watch-only node on: %s", s.rpcHost)

	return grpc.DialContext(ctx, s.rpcHost, opts...)
```

Reply:
Thanks a lot for this :rocket:! I updated `getClientConn` to include this so that it's a lot more readable :rocket:!

> I think we can simplify the readability of this method quite a bit if we add a `CancelOrQuit` helper like this to `fn`.

I added this to the `fn` package in a commit. But unfortunately I can't use this in the PR as that would require that we add a replacement of the `fn` package in the `go.mod` file to use the local version. And the local `fn` version doesn't run with `master` yet either. 
I therefore temporarily added the `CancelOrQuit` in the `remote_signer_client.go` file as well, until this is resolved.

How do we normally handle such additions to the `fn` package? Do we create a small PR that just adds the functionality, then create a new tag, and then bump to the new tag in the PR that needs the functionality?